joint_trajectory_controller/src/trajectory.cpp

```diff
@@ -17,58 +17,155 @@
+    trajectory_msgs::msg::JointTrajectoryPoint & output)
+    {
+      rclcpp::Duration duration_so_far = sample_time - time_a;
+      rclcpp::Duration duration_btwn_points = time_b - time_a;
+      double percent = duration_so_far.seconds() / duration_btwn_points.seconds();
+      percent = percent > 1.0 ? 1.0 : percent;
+      percent = percent < 0.0 ? 0.0 : percent;
+
+      output.positions.resize(state_a.positions.size());
+      for (uint i = 0; i < state_a.positions.size(); ++i) {
+        output.positions[i] =
+          state_a.positions[i] + percent * (state_b.positions[i] - state_a.positions[i]);
+      }
+
+      if (!state_a.velocities.empty() && !state_b.velocities.empty()) {
+        output.velocities.resize(state_b.velocities.size());
+        for (uint i = 0; i < state_b.velocities.size(); ++i) {
+          output.velocities[i] =
+            state_a.velocities[i] + percent * (state_b.velocities[i] - state_a.velocities[i]);
+        }
+      }
+
+      if (!state_a.accelerations.empty() && !state_b.accelerations.empty()) {
+        output.accelerations.resize(state_b.accelerations.size());
+        for (uint i = 0; i < state_b.accelerations.size(); ++i) {
+          output.accelerations[i] =
+            state_a.accelerations[i] + percent *
+            (state_b.accelerations[i] - state_a.accelerations[i]);
+        }
+      }
+    };
+
+  // current time hasn't reached traj time of the first msg yet
+  const auto & first_point_in_msg = trajectory_msg_->points[0];
+  rclcpp::Duration offset(first_point_in_msg.time_from_start.sec,
+    first_point_in_msg.time_from_start.nanosec);
+  rclcpp::Time first_point_timestamp = trajectory_start_time_ + offset;
+  if (sample_time < first_point_timestamp) {
+    rclcpp::Time t0 = time_before_traj_msg_;
+
+    linear_interpolation(
+      t0, state_before_traj_msg_, first_point_timestamp, first_point_in_msg,
+      sample_time, expected_state);
+    start_segment_itr = begin();  // no segments before the first
+    end_segment_itr = begin();
+    return true;
   }
 
   // time_from_start + trajectory time is the expected arrival time of trajectory
-  for (auto point = begin(); point != end(); ++point) {
-    auto start_time = time_add(trajectory_start_time_, point->time_from_start);
-    if (time_less_than(sample_time, start_time)) {
-      return point;
+  uint last_idx = trajectory_msg_->points.size() - 1;
+  for (uint i = 0; i < last_idx; ++i) {
+    auto & point = trajectory_msg_->points[i];
+    auto & next_point = trajectory_msg_->points[i + 1];
+
+    rclcpp::Duration t0_offset(point.time_from_start.sec, point.time_from_start.nanosec);
+    rclcpp::Duration t1_offset(next_point.time_from_start.sec, next_point.time_from_start.nanosec);
+    rclcpp::Time t0 = trajectory_start_time_ + t0_offset;
+    rclcpp::Time t1 = trajectory_start_time_ + t1_offset;
+
+    if (sample_time >= t0 && sample_time < t1) {
+      // TODO(ddengster): Find a way to add custom interpolation implementations.
```

Comment:
please open a follow-up issue and link it here

Reply:
https://github.com/ros-controls/ros2_controllers/issues/30